rust/pattern/constraint/thing_constraint.rs

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+use crate::common::error::{ErrorMessage, INVALID_CONSTRAINT_DATETIME_PRECISION};
+use crate::pattern::*;
+use crate::write_joined;
+use chrono::{NaiveDateTime, Timelike};
+use std::fmt;
+use std::fmt::{Display, Write};
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct IsaConstraint {
+    pub type_: TypeVariable,
+    pub is_explicit: bool,
+}
+
+impl<T: Into<ScopedType>> From<T> for IsaConstraint {
+    fn from(type_name: T) -> Self {
+        IsaConstraint {
+            type_: UnboundVariable::hidden().type_(type_name).unwrap().into_type(),
+            is_explicit: false,
+        }
+    }
+}
+
+impl From<UnboundVariable> for IsaConstraint {
+    fn from(var: UnboundVariable) -> Self {
+        IsaConstraint::from(var.into_type())
+    }
+}
+impl From<TypeVariable> for IsaConstraint {
+    fn from(type_: TypeVariable) -> Self {
+        IsaConstraint { type_, is_explicit: false }
+    }
+}
+
+impl Display for IsaConstraint {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "isa {}", self.type_)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct HasConstraint {
+    pub type_: Option<TypeVariable>,
+    pub attribute: ThingVariable,
+}
+
+impl HasConstraint {
+    pub fn from_value(type_name: String, value: ValueConstraint) -> Self {
+        HasConstraint {
+            type_: Some(UnboundVariable::hidden().type_(type_name).unwrap().into_type()),
+            attribute: UnboundVariable::hidden().constrain_value(value),
+        }
+    }
+
+    pub fn from_typed_variable(type_name: String, variable: ThingVariable) -> Self {
+        HasConstraint {
+            type_: Some(UnboundVariable::hidden().type_(type_name).unwrap().into_type()),
+            attribute: variable,
+        }
+    }
+
+    pub fn from_variable(variable: ThingVariable) -> Self {
+        HasConstraint { type_: None, attribute: variable }
+    }
+}
```

Comment:
This should probably be reworked into a set of `From<...>` impls for consistency. That'd also make `ThingVariableBuilder` more uniform.